### PR TITLE
feat(disk): reject write if disk space is insufficient

### PR DESCRIPTION
**Pull request by hycdong**
This pull request is about disk protection when space is almost full. If one disk available space is below `disk_min_available_space_ratio`, all replica on this disk will reject user write request. The disk status is updated periodically, when disk space is normal, user write request will be recovered.

This pull request also adds related unit tests and dynamical configuration.
```diff
[replication]
+enable_disk_available_space_check=true
+disk_min_available_space_ratio=10
```

### CODE REVIEW

File: src/client/partition_resolver.cpp [hunk 1]
```diff
@@ -53,7 +53,8 @@ void partition_resolver::call_task(const rpc_response_task_ptr &t)
         dsn::error_code err, dsn::message_ex * req, dsn::message_ex * resp)
     {
         if (req->header->gpid.value() != 0 && err != ERR_OK && err != ERR_HANDLER_NOT_FOUND &&
-            err != ERR_APP_NOT_EXIST && err != ERR_OPERATION_DISABLED && err != ERR_BUSY) {
```
Comment (levy5307) [inline src/client/partition_resolver.cpp:56]: you can put all of there error code into a contianer, and see whether it's in the container or not

File: src/common/fs_manager.cpp [hunk 2]
```diff
@@ -75,23 +89,40 @@ unsigned dir_node::remove(const gpid &pid)
     return iter->second.erase(pid);
 }
 
-void dir_node::update_disk_stat()
+void dir_node::update_disk_stat(bool &status_changed)
```
Comment (levy5307) [inline src/common/fs_manager.cpp:92]: ```bool dir_node::update_disk_stat()```

File: src/replica/test/replica_disk_test_base.h [hunk 3]
```diff
@@ -95,6 +97,32 @@ class replica_disk_test_base : public replica_test_base
         }
     }
 
+    void
+    mock_node_status(int32_t node_index, disk_status::type old_status, disk_status::type new_status)
+    {
+        auto node = get_dir_nodes()[node_index];
+        for (auto &kv : node->holding_replicas) {
```
Comment (levy5307) [inline src/replica/test/replica_disk_test_base.h:104]: const auto &

File: src/common/fs_manager.cpp (hunk 2 above)
Comment (hycdong) [reply src/common/fs_manager.cpp:92]: I used to clear function like your suggestion. If return false, it seems like update disk stat failed, so I pass `status_chaned` instead.

**Review by acelyc111 — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]

**Pull request merged**